scipy/optimize/_direct.py

```diff
@@ -0,0 +1,182 @@
+from .optimize import OptimizeResult
+
+ERROR_MESSAGES = (
+    "u[i] < l[i] for some i",
+    "maxfun is too large",
+    "Initialization failed",
+    "There was an error in the creation of the sample points",
+    "An error occured while the function was sampled",
+    "Maximum number of levels has been reached.",
+)
+
+SUCCESS_MESSAGES = (
+    "Number of function evaluations done is larger then maxfun={}",
+    "Number of iterations is equal to maxiter={}",
+    ("The best function value found is within {} percent "
+    "of the (known) global optimum"),
+    ("The volume of the hyper-rectangle with best function value found "
+     "is below volper={}"),
+    ("The volume of the hyper-rectangle with best function value found "
+     "is smaller then volper={}"),
+)
+
+def _minimize_direct(func, bounds=None, nvar=None, *args, disp=False,
+                     eps=1e-4, maxfun=20000, maxiter=6000, 
+                     locally_biased=False, fglobal=-1e100, fglper=0.01, 
+                     volper=-1.0, sigmaper=-1.0):
+    r"""
+
+    Solve an optimization problem using the DIRECT 
+    (Dividing Rectangles) algorithm.
+    It can be used to solve general nonlinear programming problems of the form:
+
+    .. math::
+
+           \min_ {x \in R^n} f(x)
+
+    subject to
+
+    .. math::
+
+           x_L \leq  x  \leq x_U
+
+    Where :math:`x` are the optimization variables (with upper and lower
+    bounds), :math:`f(x)` is the objective function.
+
+    Parameters
+    ----------
+    func : objective function
+        called as `func(x, *args)`; does not need to be defined everywhere,
+        raise an Exception where function is not defined
+    bounds : array_like, optional
+            ``(min, max)`` pairs for each element in ``x``, defining
+            the bounds on that parameter.
+    nvar : int, optional
+        Dimensionality of x (only needed if `bounds` is not defined)
+    eps : float, optional
+        Ensures sufficient decrease in function value when a new potentially
+        optimal interval is chosen.
+    maxfun : int, optional
+        Approximate upper bound on objective function evaluations.
+        Maximal allowed value is 90000 see documentation of Fortran library.
+    maxiter : int, optional
+        Maximum number of iterations.
+        Maximal allowed value is 6000 see documentation of Fortran library.
+    locally_biased : bool, optional
+        Whether to use the original or modified DIRECT algorithm.
+        Possible values:
+
+        * ``locally_based=False`` - use the original DIRECT algorithm
+        * ``locally_biased=True`` - use the modified DIRECT-l algorithm
+    fglobal : float, optional
+        Function value of the global optimum.
+        By default it is a negative value whose absolute value is very large.
+        Set this value only if the global optimum is known.
+    fglper : float, optional
+        Terminate the optimization when the percent error satisfies:
```

Comment:
Is it possible to terminate when the change in f between iterations is below a certain value (i.e. not compared to an `fglobal`)

Reply:
I don't think so, the algorithm considers function as a black box and whatever information is available is through the user inputs (such as `fglobal`) which it may utilize.